Mixins to simplify accessing SettingsService and theming

e-kolpakov opened the pull request:
*Description:* this PR contains several mixins to siplify and streamline accessing XBlock-specific settings via `SettingsService`:
* Mixin to provide access to settings service
* Mixin for configurable themes
* Tests for both

*JIRA Ticket:* [OSPR-582](https://openedx.atlassian.net/browse/OSPR-582)
*Related:* [cleaning up Poll XBlock for edx.org](https://github.com/mckinseyacademy/xblock-poll/pull/13)
*Test instructions:* No user facing changes - run test suite.

e-kolpakov commented:
@sarina just a little reminder - this waits for your review and blocks further Poll & Survey cleanup

sarina commented on xblockutils/settings.py:
```diff
@@ -0,0 +1,87 @@
+    default_theme_config = None
+    theme_key = "theme"
+
+    def get_theme(self):
```
Is theme support a new thing? I guess I haven't heard about getting theme specific stuff from the XBlock itself and thus am a bit confused how this works. Is this something the xblock runtime natively supports and you're just making access easier, or is this a new thing?

e-kolpakov replied:
@sarina TL;DR basically this is XBlock-specific theme, not LMS/Studio theme. It is limited to CSS files and basically only changes look and feel of an XBlock. Runtime provides more generic `SettingsService` which can be used for theming purposes (and actually used in mentoring XBlock)

it's old new thing. There exists [`SettingsService`](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/services.py) allowing XBlocks to be configurable per environment (i.e. via [cl]ms.env.json). The only usage I'm aware of as of now is mentoring (aka problem builder) [theming](https://github.com/open-craft/problem-builder/blob/master/problem_builder/mentoring.py#L214-225).

So, since Poll and Survey need theming as well it was natural to reuse this approach. In order to avoid code duplication I basically extracted theming pieces from mentoring into xblock-utils. There's a PR that switches mentoring to use this code as well: https://github.com/open-craft/problem-builder/pull/31.

sarina commented on xblockutils/settings.py:
```diff
@@ -0,0 +1,87 @@
+
+    Examples:
+
+        Looks up red.css and small.css in `my_xblock` package:
```
So these .css files would live within your XBlock?

e-kolpakov replied:
@sarina initial use case was to host theme css in XBlock. However it is flexible and allows hosting them in arbitrary package.

Basically, theme config have two parts: 

* `package` - "top-level" selector specifying where to look for theme files
* `locations` - a list of locations within that package

sarina replied:
Is there documentation on how to take advantage of this, beyond docstrings?

sarina commented on xblockutils/settings.py:
```diff
@@ -0,0 +1,87 @@
+            'locations': ['red.css', 'small.css']
+        }
+
+        Looks up public/themes/red.css in my_other_xblock.assets
```
These .css files live where? WIthin another block? Within the host platform?

e-kolpakov replied:
@sarina they should be package resources. Mentoring provides [two themes](https://github.com/open-craft/problem-builder/tree/master/problem_builder/public/themes). They are listed in [mentoring's setup.py](https://github.com/open-craft/problem-builder/blob/master/setup.py#L88) (implicitly, basically `public` folder is included as a package resource and themes live in nested folder) and can be accessed using the following theme configurations:

    {
        'package': 'problem_builder',
        'locations': ['public/themes/lms.css']
    }

or

    {
        'package': 'problem_builder',
        'locations': ['public/themes/apros.css']
    }

The idea was to allow to manage themes and XBlock itself separately, so that someone who just wants to provide a custom theme for an XBlock wouldn't have to fork it and maintain the fork. All he has to do is to create a package with theme, install it into LMS environment and configure XBlock to use the package as a theme source.

sarina commented:
@e-kolpakov this code looks good. Before you merge I just want a little clarity on how this is used,etc.

sarina commented:
@e-kolpakov I'm :+1: with this. I'd like there to be a bit of documentation on its use, possibly within this repo's README file or github wiki (or both - README could point it out and direct to a wiki page for further discussion). Particularly I'm interested to let others know how they can take advantage of using this to package their own theme.

I don't consider this a blocker to merging, but would like to make sure it's done. This will also help us if we need to trace through an aspect of your block that is involving theming in the future.

e-kolpakov commented:
@sarina ok, I've added theming and settings sections into README (linking to wiki page, thank you for the suggestion). Also, while creating wiki page I've identified and implemented a small improvement - just a bit of checks to throw exceptions less. Not a big deal + they are covered with new tests, so I'm not requesting a review on them. Since there are significant a11y notes on Poll it's not urgent to merge this PR, so I'll merge it tomorrow morning to give you a chance to review code changes and docs, if you want to.

sarina commented:
:+1: thanks @e-kolpakov looks great
